WIP compare diagnoses

graemeblair opened the pull request:
opening PR for code review

lilymedina committed:
compare_diagnoses WIP

lilymedina committed:
compare_diagnoses methods wip

lilymedina committed:
minor correction

lilymedina committed:
corrections

lilymedina committed:
## comment class

lilymedina committed:
print method partially working

lilymedina committed:
add checks

Commit:
wip compare_diagnoses

not working as it is.

Commit:
prepare data.frame for output

Commit:
compare_diagnoses looking good

Commit:
compare_diagnoses improved

Commit:
document

lilymedina committed:
Document Function and clean up code

lilymedina committed:
printe output looking better

lilymedina committed:
minor edits

lilymedina committed:
fix bug

lilymedina committed:
document and print function

lilymedina committed:
Minor edits to documentation

lilymedina committed:
edit args names

lilymedina committed:
clean documentation

lilymedina committed:
95% typo

lilymedina committed:
# temp edits

lilymedina committed:
bug

lilymedina committed:
print 2 digits numbers

lilymedina committed:
default merge by estimator to TRUE

lilymedina committed:
fix print bug

lilymedina committed:
Fix bug print function

macartan committed:
defaults and bits and pieces

lilymedina committed:
add confidence level as arg

lilymedina committed:
update compare_diagnoses

Clean up code

macartan committed:
alpha

graemeblair committed:
Merge branch 'master' into compare_diagnoses

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,345 @@
+                              merge_by_estimator = TRUE){
+  
+  
+  if(bootstrap_sims <= 99) stop("Please choose a higher number of bootstrap simulations")
```
In `diagnose_design` this is a warning and uses a different number. Should they be consistent?

lilymedina replied:
The function - as it currently is - checks which of the diagnosands of the`comparison_design` lie outside the (1-alpha)% bootstrap confidence interval of their counterparts in the `base_design` . So we do need bootstrap replicates (at least for the first design), and a high number of those (and sims) to gain precision and make compare_diagnoses stable.

graemeblair replied:
Makes sense. Can the default # be the same? Right now it's 30 in `simulate_design` and 99 here.

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,345 @@
+                                  sims = sims, 
+                                  bootstrap_sims = bootstrap_sims)
+
+    } else if(class(base_design) == "diagnosis") {
```
can you reindent all the code? a bit hard to follow. CMD+SHIFT+A in RStudio on the whole file will do the trick.

nfultz replied:
> CMD+SHIFT+A in RStudio on the whole file will do the trick.

Please do not do this - try to follow either the google style or hadley's. Or use Yihui's package - https://yihui.name/formatr/

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,345 @@
+
+    } else if(class(base_design) == "diagnosis") {
+      if(base_design$bootstrap_sims<= 99)
+        stop("base_design must have a higher number of bootstrap simulations")
```
error should report that it must be above 99

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,345 @@
+    } else if(class(base_design) == "diagnosis") {
+      if(base_design$bootstrap_sims<= 99)
+        stop("base_design must have a higher number of bootstrap simulations")
+    diagnosis1 <- base_design} else{ 
```
clean up spacing and line breaks

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,345 @@
+    
+    diagnosis2 <- diagnose_design(comparison_design,
+                                  sims = sims, 
+                                  bootstrap_sims = 100)
```
why is 100 hard-coded here? should be an option

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,345 @@
+                              comparison_design, 
+                              sims = 1000,
+                              bootstrap_sims = 1000, 
+                              alpha = 0.01,
```
why .01 not .05?

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,345 @@
+compare_diagnoses <- function(base_design,
+                              comparison_design, 
+                              sims = 1000,
+                              bootstrap_sims = 1000, 
```
can these defaults be the same as `diagnose_design` to reduce confusion?

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,345 @@
+    } else if(class(comparison_design) == "diagnosis"){
+      diagnosis2 <- comparison_design
+      } else{ 
+        stop("comparison_design must be either a design or a diagnosis")}
```
spacing

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,345 @@
+  
+  # 1.Housekeeping
+  if(is.null(diagnosis1$bootstrap_replicates ))
+    stop("Can't compare diagnoses witouth bootstrap replicates")
```
this should be a more informative error, giving help for how to include bootstrap, i.e. "please set bootstrap_sims to a number above zero". also spelling.

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,345 @@
+  if(is.null(diagnosis1$bootstrap_replicates ))
+    stop("Can't compare diagnoses witouth bootstrap replicates")
+  
+  if(length(diagnosis1$parameters_d[, "design_label"])  + length(diagnosis2$parameters_d[, "design_label"])> 2) 
```
will this work in general if there is not a parameters_df?

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,345 @@
+    stop("Please only send design or diagnosis objects with one unique design_label.")
+
+  
+  diagnosands  <- base::intersect(diagnosis1$diagnosand_names, 
```
call this diagnosand_names

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,345 @@
+  group_by_set1 <- diagnosis1$group_by_set
+  group_by_set2 <- diagnosis2$group_by_set
+  
+  mand  <-  "estimand_label"  %in% group_by_set1 & "estimand_label"  %in% group_by_set2
```
please use full words here, following code style in rest of package

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,345 @@
+  
+  mand  <-  "estimand_label"  %in% group_by_set1 & "estimand_label"  %in% group_by_set2
+  mator <-  "estimator_label" %in% group_by_set1 & "estimator_label" %in% group_by_set2
+  if( mand + mator < 1) 
```
better would be to check what group_by_set1 and group_by_set2 have in common and then check to see if it includes estimand label and estimator label or both

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,345 @@
+  mand  <-  "estimand_label"  %in% group_by_set1 & "estimand_label"  %in% group_by_set2
+  mator <-  "estimator_label" %in% group_by_set1 & "estimator_label" %in% group_by_set2
+  if( mand + mator < 1) 
+    stop(paste0("diagnosands_df must contain at least estimand_label or estimator_label in common" ))
```
don't need paste0. clean up text.

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,345 @@
+  
+  merge_by_set <- NULL
+
+  if(mand) merge_by_set <- c("estimand_label")
```
please use curly braces for if's and no need for c()

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,345 @@
+  if(mand) merge_by_set <- c("estimand_label")
+  if(merge_by_estimator){
+    if(!mator)
+      warning("Estimator label not used in columns for merging.
```
one line

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,345 @@
+
+  if(mand) merge_by_set <- c("estimand_label")
+  if(merge_by_estimator){
+    if(!mator)
```
mator == FALSE

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,345 @@
+    if(!mator)
+      warning("Estimator label not used in columns for merging.
+              At least one of the designs doesn't contain estimator_label.")
+    else
```
curly braces

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,345 @@
+   merge_by_set <- c(merge_by_set, "estimator_label")
+  }
+  
+  if("term" %in% group_by_set1 & "term" %in% group_by_set2) 
```
curly braces; clean up spacing

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,345 @@
+  
+  if(nrow(comparison_df) == 0 ){
+    warning("Can't merge diagnosands data frames. Diagnoses don't have labels in common")
+   return(list(diagnosis1 = diagnosis1 , diagnosis2 = diagnosis2))
```
should just error, not return diagnoses -- the user already has access to both diagnoses -- and output should be predictable

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,345 @@
+    
+  c_names  <- colnames(comparison_df)
+  suffix   <- c("_1", "_2")
+  dropcols <- grepl("[[:digit:]]+$", c_names ) |   c_names  %in% c("design_label", "estimand_label", "estimator_label", "term") 
```
is there a cleaner way to do this? seems fragile

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,345 @@
+  # just in case diagnosis2 doesn't  include bootrstrap_replicates
+  filter_se      <- grepl("^se",  c_names)
+  diagnosands_se <-  comparison_df[, filter_se]
+  if(diagnosis2$bootstrap_sims == 0){
```
spacing

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,345 @@
+  filter_se      <- grepl("^se",  c_names)
+  diagnosands_se <-  comparison_df[, filter_se]
+  if(diagnosis2$bootstrap_sims == 0){
+  diagnosands_se1 <- diagnosands_se; diagnosands_se2 <- matrix(data = rep(NA, length(diagnosands_se)))
```
please line break and don't use semicolons

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,345 @@
+  labels_df     <- lapply(labels_df, head, n = 1)
+  bootstrap_df  <- split(bootstrap_df , lapply(group_by_list, addNA), drop = TRUE)
+  group_by_set  <- diagnosis1$group_by_set
+  lower.bound   <- lapply(bootstrap_df,FUN = function(x) {
```
lower.bound -> lower_bound

lower_bound and upper_bound should be the same function but use an argument for which you should return. they are identical except the quantile.

please do not use objects from the global environment within the function, include them as additional arguments to lapply (such as alpha and group_by_set)

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,345 @@
+  filter_mean      <- gsub("_[[:digit:]]+$","", colnames(comparison_df)) %in%  diagnosands
+  diagnosands_mean <- comparison_df[,  filter_mean]
+  filter_sims      <- grepl("^n_sims", colnames(comparison_df))
+  sims_df          <- comparison_df[,filter_sims]
```
drop = FALSE?

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,345 @@
+                  #args
+                  mean_1 = diagnosands_mean[pair, 1:length(diagnosands)],
+                  mean_2 = diagnosands_mean[pair, (length(diagnosands)+1):ncol(diagnosands_mean)],
+                  se_1 = diagnosands_se1[pair,],
```
drop = FALSE here and throughout?

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,345 @@
+  c_names <- colnames(out)
+  
+  #comparison_df <-  rapply(object = out , f = round, classes = "numeric", how = "replace", digits = 2) 
+  comparison_df <- out
```
would prefer not to be renaming objects, can we keep it as a single name?

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,345 @@
+  
+  # Prepare output
+  # Add "_1" or "_2" to labels as needed.
+  if("estimand_label" %in% c_names & ! "estimand_label" %in% merge_by_set)
```
curly braces; spacing

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,345 @@
+              diagnosis1 = diagnosis1,
+              diagnosis2 = diagnosis2)
+
+  class(out) <- "compared.diagnoses"
```
diagnosis_comparison

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,345 @@
+  n_sims1 <- nrow(x$diagnosis1$simulations_df)
+  n_sims2 <- nrow(x$diagnosis2$simulations_df)
+  x  <- x[["compared.diagnoses_df"]]
+  sx <- subset(x,  x[,"in_interval"] == 0)
```
please use more informative names

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,345 @@
+    cat(paste0("\n Comparison of research designs diagnoses based on ", n_sims1, " simulations", atext))
+  else 
+    cat(paste0("\n Comparison of research designs diagnoses based on ", n_sims1, " simulations from `base_design` and ", n_sims2, " from `comparison_design`", atext))
+  if(bootstrap_rep1 == bootstrap_rep2)
```
curly braces throughout here

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,345 @@
+
+  
+  # Prep vars to shape data frame
+  mand  <- startsWith(colnames(sx), "estimand")
```
please use longer more informative names

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,345 @@
+  diagnosands_m <- sapply(sx[,diagnosands_cols], format_num, digits=2)
+  
+  
+  out[odd, diagnosands_cols ]  <- diagnosands_m
```
please follow reshape_diagnosis behavior here, this is a whole new set of behaviors and would be good to use the same to minimize code maintenance going forward

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,345 @@
+    
+  cat("\n\n")
+  print(out, row.names = FALSE)
+  cat(paste0("\n\n Displaying diagnosands that statistically diverge between base and comparison designs. See help file for details." ))
```
Shouldn't send to help file, should communicate the test used.

graemeblair reviewed (commented):
Thanks for this great work!

Lots of small style comments throughout, please excuse brevity!

Bigger suggestions:
1) add difference and standard error of difference via sqrt( se_1^2 + se_2^2 )
2) by default bootstrap both if designs are sent, so that the se of diff can be calculated

lilymedina committed:
Update compare_diagnoses.R

 - reindent code
- make print function similar to reshape_diagnosis
- spell out object names

lilymedina committed:
get rid of line breaks in if stop

lilymedina commented:
Thanks for reviewing. All small comments have been addressed. 

Re difference sounds good. Although I'm just not sure if it suggests a test of difference in means that we're not using.--I'll add columns anyway. 

**TODO:**
- [x] write tests
- [ ] add difference and se of difference columns

lilymedina committed:
create bootstrap arg

lilymedina committed:
document compare diagnoses internal

lilymedina committed:
update argument names

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,331 @@
+  
+ # Diagnose designs if base/comparison are of class "design"
+  
+  if( "design" %in% class(base_design)  ){
```
Please use check_design_class_single(base_design) here

lilymedina replied:
Ignoring this suggestion since function should also work for objects of classe `diagnosis`

nfultz replied:
Also you can use the `inherits` function instead.

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,331 @@
+    diagnosis1 <- base_design} else stop("base_design must be either a design or a diagnosis")
+  
+
+  if( "design" %in% class(comparison_design)  ){
```
check_design_class_single

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,331 @@
+
+    } else if(class(base_design) == "diagnosis") {
+      if(base_design$bootstrap_sims<= 99) stop("base_design must have a higher number of bootstrap simulations")
+    diagnosis1 <- base_design} else stop("base_design must be either a design or a diagnosis")
```
please add line breaks around the } -- very hard to read

lilymedina replied:
so sorry!
Hope is ok now.

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,331 @@
+      diagnosis2 <- comparison_design else stop("comparison_design must be either a design or a diagnosis")
+  
+  
+  out <- compare_diagnoses_internal(diagnosis1, diagnosis2, merge_by_estimator, alpha )
```
please fix indenting in this area

graemeblair commented on R/compare_diagnoses.R:
```diff
@@ -0,0 +1,331 @@
+  if (length(diagnosis1$parameters_d[, "design_label"]) + length(diagnosis2$parameters_d[, "design_label"]) > 2) stop("Please only send design or diagnosis objects with one unique design_label.")
+  
+  
+  diagnosands <- base::intersect(diagnosis1$diagnosand_names, diagnosis2$diagnosand_names)
```
is base:: necessary here?

graemeblair commented:
Just checking in on this. A couple of things still need to be addressed, including:
1) cleaning up spacing and indentation to make the code more readable;
2) using curly braces around all ifs; 
3) simplifying lower.bound and upper.bound to a single function that is used twice (the code is identical except one number);
4) not exporting the internal function.

Let us know when you'd like us to comment again! Added a few but stopped because didn't want to interrupt the work in progress.

lilymedina committed:
indentation, tests and messaging

-

lilymedina committed:
merge master

lilymedina committed:
solve rd conflicts

lilymedina committed:
add more tests

Commit:
Increase coverage and change class name

Commit:
add difference columns and name w underscore

lilymedina commented:
Thanks for the review. Hoping code is easier to read now. 

Main updates are tests and mean and se difference columns; se difference is computed and returned if the `comparison_design` has bootstrap sims -- I still think these columns can be misleading as they might make people think that we are doing a mean comparison test.


Feel free to review and edit functions as needed so that it can be integrate with compare_designs.

lilymedina committed:
stop travis from breaking

lilymedina committed:
travis not passng test?

graemeblair committed:
Merge branch 'master' into compare_diagnoses

graemeblair committed:
R CMD CHECK changes

graemeblair committed:
renaming and code cleanup

acoppock commented:
Hi all -- we took a spin through this today prepping for a small DD update. Can I request 3 changes?

1. I would prefer it if the output did *not* subset to those rows where the diagnosands are "out of interval". This changes quite a bit from sim to sim, and in any case, I'd like to *see* that the diagnosands are similar, not just have them subset out.

2. Right now it appears that we show the estimate and standard error for the first design, but only the estimate for the second design.  Could we please show the estimate and standard error for both?

3. GB requested the difference and the standard error of the difference be added as a third column.  I think this is crucial, since that comparison is the main quantity of interest. I think that this should always be included. [I see a comment above that this could be toggled on and off, but I think it should always be displayed -- I mean, *aren't* we doing a pseudo mean comparison test with this confidence interval approach? Maybe I'm not following, sorry about that!]

Thank you!

Commit:
difference column
